### PR TITLE
Migrate precompiles, opcodes, eei, message, txContext to typescript

### PR DESCRIPTION
**Pull request by s1na**
Happy for now with precompiles, but the rest are still subject to change. E.g. `Message`'s constructor gets an `opts: any`, which doesn't utilize typescript's benefits.

I also tried migrating `Interpreter` and `Loop` but realized halfway that they need some changes before they can be typed well.

(To be merged into #479)

### CODE REVIEW

**Comment by holgerd77** [comment]
Phew. 😄

**Comment by s1na** [comment]
Sorry didn't expect the diff to become so large :see_no_evil: 

Might be easier to go through these commits:

- https://github.com/ethereumjs/ethereumjs-vm/pull/497/commits/6d3df3f096f5b2a6e7acefad67c1cd1b29816887
- https://github.com/ethereumjs/ethereumjs-vm/pull/497/commits/3c7b58d7b2d8fae3b74bed53e7a0495305afc757
- https://github.com/ethereumjs/ethereumjs-vm/pull/497/commits/55fbb2ea97007bee25edb9dc9c41a76fcc71ac7c
- https://github.com/ethereumjs/ethereumjs-vm/pull/497/commits/2a63847f9319e323b23c1eea25c97ea32ca6d6fc

Btw, the large commit history is because I'm reusing the same branch for the last couple of PRs.

**Comment by holgerd77** [comment]
Ah, so this is actually ready for review? Misunderstood first.

**Comment by holgerd77** [comment]
Just scrolled through for a first round, think this will be manageable for review. 😀😛

File: lib/evm/eei.ts [hunk 1]
```diff
@@ -530,16 +534,16 @@ module.exports = class EEI {
...
-function trap (err) {
+function trap (err: string) {
   throw new VmError(err)
 }
 
 const MASK_160 = new BN(1).shln(160).subn(1)
-function addressToBuffer (address) {
+function addressToBuffer (address: BN) {
   return address.and(MASK_160).toArrayLike(Buffer, 'be', 20)
 }
```
Comment (holgerd77) [inline lib/evm/eei.ts]: Looks good (``eei.ts``).

File: lib/evm/message.ts [hunk 2]
```diff
@@ -17,7 +31,7 @@ module.exports = class Message {
     this.delegatecall = opts.delegatecall || false
   }
 
-  get codeAddress () {
+  get codeAddress (): Buffer {
     return this._codeAddress ? this._codeAddress : this.to
   }
 }
```
Comment (holgerd77) [inline lib/evm/message.ts]: Ok (``message.ts``).

File: lib/evm/opcodes.ts [hunk 3]
```diff
@@ -1,4 +1,12 @@
-const codes = {
+export interface OpInfo {
+  name: string;
+  opcode: number;
+  fee: number;
+  dynamic: boolean;
+  isAsync: boolean;
+}
```
Comment (holgerd77) [inline lib/evm/opcodes.ts]: Nice. 😄

File: lib/evm/opcodes.ts [hunk 4]
```diff
@@ -192,5 +200,5 @@ module.exports = function (op, full, freeLogs) {
     }
   }
 
-  return {name: opcode, opcode: op, fee: fee, dynamic: code[2], async: code[3]}
+  return { name: opcode, opcode: op, fee: fee, dynamic: code[2], isAsync: code[3] }
 }
```
Comment (holgerd77) [inline lib/evm/opcodes.ts]: Ok (``opcodes.ts``).

File: lib/evm/precompiles/types.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+export function OOGResult(gasLimit: BN): PrecompileResult {
+  return {
+    return: Buffer.alloc(0),
+    gasUsed: gasLimit,
+    exception: 0, // 0 means VM fail (in this case because of OOG)
+    exceptionError: ERROR.OUT_OF_GAS
+  }
+}
```
Comment (holgerd77) [inline lib/evm/precompiles/types.ts]: Really really nice to have these new precompile interfaces and to have this more systematically structured and wrapped like this.

File: lib/evm/precompiles/01-ecrecover.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+      exception: 1
+    }
+  }
+
+  return {
+    gasUsed,
+    return: setLengthLeft(publicToAddress(publicKey), 32),
+    exception: 1
+  }
+}
```
Comment (holgerd77) [inline lib/evm/precompiles/01-ecrecover.ts]: Ok, really like this results removal - a lot cleaner and better to read - as well as the dedicated ``OOGResult`` function.

File: lib/evm/precompiles/02-sha256.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+  if (opts.gasLimit.lt(gasUsed)) {
+    return OOGResult(opts.gasLimit)
+  }
+
+  return {
+    gasUsed,
+    return: sha256(data),
+    exception: 1
+  }
+}
```
Comment (holgerd77) [inline lib/evm/precompiles/02-sha256.ts]: Ok.

File: lib/evm/precompiles/03-ripemd160.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  if (opts.gasLimit.lt(gasUsed)) {
+    return OOGResult(opts.gasLimit)
+  }
+
+  return {
+    gasUsed,
+    return: ripemd160(data, true),
+    exception: 1
+  }
+}
```
Comment (holgerd77) [inline lib/evm/precompiles/03-ripemd160.ts]: Ok.

File: lib/evm/precompiles/04-identity.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+  if (opts.gasLimit.lt(gasUsed)) {
+    return OOGResult(opts.gasLimit)
+  }
+
+  return {
+    gasUsed,
+    return: data,
+    exception: 1
+  }
+}
```
Comment (holgerd77) [inline lib/evm/precompiles/04-identity.ts]: Ok.

File: lib/evm/precompiles/05-modexp.ts [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+  } else {
+    R = expmod(B, E, M)
+  }
+
+  return {
+    gasUsed,
+    return: R.toArrayLike(Buffer, 'be', mLen.toNumber()),
+    exception: 1
+  }
+}
```
Comment (holgerd77) [inline lib/evm/precompiles/05-modexp.ts]: Ok.

File: lib/evm/precompiles/06-ecadd.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  if (returnData.length !== 64) {
+    return OOGResult(opts.gasLimit)
+  }
+
+  return {
+    gasUsed,
+    return: returnData,
+    exception: 1
+  }
+}
```
Comment (holgerd77) [inline lib/evm/precompiles/06-ecadd.ts]: Ok.

File: lib/evm/precompiles/07-ecmul.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  if (returnData.length !== 64) {
+    return OOGResult(opts.gasLimit)
+  }
+
+  return {
+    gasUsed,
+    return: returnData,
+    exception: 1
+  }
+}
```
Comment (holgerd77) [inline lib/evm/precompiles/07-ecmul.ts]: Ok.

File: lib/evm/precompiles/08-ecpairing.ts [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+  if (returnData.length !== 32) {
+    return OOGResult(opts.gasLimit)
+  }
+
+  return {
+    gasUsed,
+    return: returnData,
+    exception: 1
+  }
+}
```
Comment (holgerd77) [inline lib/evm/precompiles/08-ecpairing.ts]: Ok.

**Pull request merged**